### DISCUSSION TITLE
Fetching ligature properties from Harfbuzz

### DISCUSSION THREAD

**Discussion by DeependraKhetwal** [harfbuzz/harfbuzz#4736] [General] [answered]
I want to know the ligature glyph properties ( for example the number of ligature components) for our use case . Currently i don't see any glyphInfo public API for getting it . I see that harfbuzz sets this info internally in variable `var1.u8[2]`  ( https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-layout.hh#L146) . I am able to get the number of ligature components by doing something like info[i]->var1.u8[2] & 0x0F; (for ith glyph similar in the way accessed by harfbuzz internally in `_hb_glyph_info_get_lig_num_comps`). Is it okay and safe to use these internal variables as such? If not what is the alternative way to access these properties?

**Comment by behdad** [comment]
The internals are IMO too much info to expose. You can, with fonts that have it, fetch the number of ligature components using `hb_ot_layout_get_ligature_carets`, but that would not be most fonts I think.

What's your use case?

**Reply by DeependraKhetwal** [reply]
In our text composer we use a legacy Line shaping engine which we are currently in process of replacing by harfbubzz. That legacy Line shaper provides our text composer various glyph processing information recorded during shaping process (eg how ligature processing was done using which components , number of components , which part of ligature glyph corresponds to which unicode etc) . This glyph processing data we are unable to get using harfbuzz ( this data is different and more detailed than the data we get using clusters). This data is used by our text composer at many places to do correct layout. Currently we are not able to do tab alignment properly without this information because our text composer uses the number of ligature components (provided to it by legacy Line shaper) for tab position calculations. It is okay if harfbuzz does not provide any public APIs for these . We can access them with something like `info[i]->var1.u8[2] & 0x0F;`  . We just want to make sure that these variables are safe to use from a future standpoint ( they will not be modified or used for something else in future)

**Reply by behdad** [reply]
> We just want to make sure that these variables are safe to use from a future standpoint ( they will not be modified or used for something else in future)

Exactly why we don't expose those. It's unlikely, but possible, that they change. You'd be on your own.

As for alternative approaches, I suggest trying to count the Unicode graphemes in the cluster instead.

**Reply by DeependraKhetwal** [reply]
Our Legacy line shaping engine uses the ligature components from the font so getting ligature components like above is a better approach for us than counting Unicode graphemes. While we understand that harfbuzz cannot guarantee changing internals in future we would request you to keep these kind of glyph properties unchangeable if possible in future versions so that users with use cases like us can utilize them.

**Reply by behdad** [reply, marked as answer]
Thanks. I'll think about making some of those public.
